salt-mode.el

```diff
@@ -194,6 +213,202 @@ module to use.
+
+(defconst salt-mode--false-string
+  (propertize "false" 'face 'font-lock-constant-face)
+  "String to use for showing false values.")
+
+(defconst salt-mode--true-string
+  (propertize "true" 'face 'font-lock-constant-face)
+  "String to use for showing true values.")
+
+(defconst salt-mode--null-string
+  (propertize "null" 'face 'font-lock-constant-face)
+  "String to use for showing null values.")
+
+(defconst salt-mode--mandatory-string
+  (propertize "?" 'face 'font-lock-keyword-face)
+  "String to use for showing true values.")
+
+(defconst salt-mode--kwargs-string
+  (propertize "... " 'face 'font-lock-keyword-face)
+  "String to use for unknown keyword arguments.")
+
+(defun salt-mode--format-argspec (name default)
+  "Format argument NAME with value DEFAULT for display."
+  (let ((name (propertize name 'face 'font-lock-variable-name-face)))
+    (if (member default salt-mode-hide-eldoc-argument-values)
+        name
+      (format "%s: %s" name
+              (cond ((eq default :json-false) salt-mode--false-string)
+                    ((eq default t) salt-mode--true-string)
+                    ((eq default :mandatory) salt-mode--mandatory-string)
+                    ((not default) salt-mode--null-string)
+                    (t (propertize
+                        (prin1-to-string default)
+                        'face 'font-lock-string-face)))))))
+
+(defun salt-mode--eldoc ()
+  "ElDoc support for salt-mode."
+  (when salt-mode--state-argspecs
+    (let* ((state-function (thing-at-point 'salt-mode-state-function))
+           (argspec (gethash state-function salt-mode--state-argspecs)))
+      (when argspec
+        (let* ((args (gethash "args" argspec))
+               (defaults (gethash "defaults" argspec))
+               (all-defaults
+                (append (make-list (- (length args) (length defaults)) :mandatory)
+                        defaults)))
+          (format "%s: [ %s %s]"
+                  state-function
+                  (string-join
+                   (cl-mapcar #'salt-mode--format-argspec
+                              (cdr args)
+                              (cdr all-defaults))
+                   ", ")
+                  (if (gethash "kwargs" argspec)
+                      salt-mode--kwargs-string "")))))))
+
+(defvar salt-mode--doc-mode-map
+  (let ((map (make-sparse-keymap)))
+    (set-keymap-parent map special-mode-map)
+    (define-key map "g" nil)
+    map))
+
+(define-derived-mode salt-mode--doc-mode special-mode "SaltStack Doc "
+  "This mode is used to display SaltStack documentation."
+  (font-lock-add-keywords nil rst-font-lock-keywords)
+  (read-only-mode t))
+
+(put #'salt-mode--doc-mode 'mode-class 'special)
+
+(defun salt-mode-describe-state (module-or-function)
+  "Show documentation for the given state MODULE-OR-FUNCTION.
+
+This command requires Salt be installed (not necessarily running)
+on the same machine as the file being edited."
+  (interactive
```

Comment:
Can you reuse `salt-mode--doc-read-arg` for getting the arguments to `(interactive)`, and add the hash-table for completion values?

I like the way `salt-mode-browse-doc` works just now - when called interactively, it will use the current state module function, but use completing-read (with a good default) when called with a prefix argument. I'd like this function to behave the same way.